pkg/controller/vm_controller.go

```diff
@@ -338,6 +357,89 @@ func (r *VMReconciler) buildVMPod(ctx context.Context, vm *virtv1alpha1.VirtualM
 				Args:            []string{volumeMount.MountPath + "/rootfs.raw", strconv.FormatInt(volume.ContainerRootfs.Size.Value(), 10)},
 				VolumeMounts:    []corev1.VolumeMount{volumeMount},
 			})
+		case volume.PersistentVolumeClaim != nil:
+			vmPod.Spec.Volumes = append(vmPod.Spec.Volumes, corev1.Volume{
+				Name: volume.Name,
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: volume.PersistentVolumeClaim.ClaimName,
+					},
+				},
+			})
+
+			pvcKey := types.NamespacedName{
+				Namespace: vm.Namespace,
+				Name:      volume.PersistentVolumeClaim.ClaimName,
+			}
+			var pvc corev1.PersistentVolumeClaim
+			if err := r.Client.Get(ctx, pvcKey, &pvc); err != nil {
+				return nil, fmt.Errorf("get volume PVC: %s", err)
+			}
+
+			volumeMode := corev1.PersistentVolumeFilesystem
+			if pvc.Spec.VolumeMode != nil {
+				volumeMode = *pvc.Spec.VolumeMode
+			}
+
+			if volumeMode == corev1.PersistentVolumeFilesystem {
+				// TODO: filesystem PVC is supported by CDI now, do we need support it directly?
+				return nil, fmt.Errorf("don't support filesystem PVC now: %s", volume.Name)
+			} else {
+				volumeDevice := corev1.VolumeDevice{
+					Name:       volume.Name,
+					DevicePath: "/mnt/" + volume.Name,
+				}
+				vmPod.Spec.Containers[0].VolumeDevices = append(vmPod.Spec.Containers[0].VolumeDevices, volumeDevice)
+			}
+		case volume.DataVolume != nil:
+			var dv cdiv1beta1.DataVolume
+			dvKey := types.NamespacedName{
+				Name:      volume.DataVolume.Name,
+				Namespace: vm.Namespace,
+			}
+			if err := r.Client.Get(ctx, dvKey, &dv); err != nil {
+				return nil, err
+			}
+			if dv.Status.Phase != cdiv1beta1.Succeeded {
+				klog.InfoS("Datavolume is not ready", "volume", volume.Name)
+				return nil, &requeueError{RequeueAfter: 5 * time.Second}
```

Comment:
It could take quite a long time for CDI to import the whole data volume, so I'm thinking it's good enough to leave it to exponential backoff for now

Reply:
return a error would disturb logs, especially when  a log collection tool is deployed. or we can make requeueAfter longer for  30 seconds? or watch DataVolume to trigger vm retry

Reply:
If we really want to requeue, a better way to do it is to return error as normal (like "fmt.Errorf("data volume is not ready")"), and then catch this error in reconcile func.

Reply:
I will remove it to use default backoff